flink-libraries/flink-state-processing-api/src/main/java/org/apache/flink/state/api/WritableSavepoint.java

```diff
@@ -88,40 +106,119 @@ public final void write(String path) {
 		List<BootstrapTransformationWithID<?>> newOperatorTransformations = metadata.getNewOperators();
 		DataSet<OperatorState> newOperatorStates = writeOperatorStates(newOperatorTransformations, savepointPath);
 
-		List<OperatorState> existingOperators = metadata.getExistingOperators();
+		List<OperatorState> existingOperatorStateList = metadata.getExistingOperators();
 
-		DataSet<OperatorState> finalOperatorStates = unionOperatorStates(newOperatorStates, existingOperators);
+		DataSet<OperatorState> finalOperatorStates;
+		if (existingOperatorStateList.isEmpty()) {
+			finalOperatorStates = newOperatorStates;
+		} else {
+			DataSet<OperatorState> existingOperatorStates = newOperatorStates.getExecutionEnvironment()
+				.fromCollection(existingOperatorStateList);
+
+			existingOperatorStates
+				.map(new OperatorState2PathSetMapFunction())
+				.flatMap(new FlatMapFunction<Set<Optional<Path>>, Optional<Path>>() {
+					@Override
+					public void flatMap(Set<Optional<Path>> value, Collector<Optional<Path>> out) throws Exception {
+						for (Optional<Path> path : value) {
+							out.collect(path);
+						}
+					}
+				})
+				.filter(Optional<Path>::isPresent)
+				.map(Optional<Path>::get)
+				.returns(TypeInformation.of(new TypeHint<Path>(){}))
+				.map(new MapFunction<Path, Path>() {
+					@Override
+					public Path map(Path existingStateFile) throws IOException {
+						savepointPath.getFileSystem().mkdirs(savepointPath);
```

Comment:
If you make this a `RichMapFunction` then this operation can be completed inside of open, instead of on every element.

Reply:
Also probably only have one subtask perform the operation. 

```java
@Override
public void open(Configuration configuratio) throws Exception {
    if (getRuntimeContext.getIndexOfThisSubtask() == 0) {
         savepointPath.getFileSystem().mkdirs(savepointPath);
    } 
}
```

Reply:
Good idea. Thanks.